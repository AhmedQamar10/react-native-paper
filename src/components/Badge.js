/* @flow */

import * as React from 'react';
import { Animated, View, StyleSheet } from 'react-native';
import color from 'color';
import Text from './Typography/Text';
import { black, white } from '../styles/colors';
import { withTheme } from '../core/theming';
import type { Theme } from '../types';

type Props = {|
  /**
   * Value of the `Badge`.
   */
  value?: string | number,
  /**
   * Vertical position of `Badge`.
   */
  verticalPosition?: 'top' | 'bottom',
  /**
   * Horizontal position of `Badge`.
   */
  horizontalPosition?: 'left' | 'right',
  /**
   * Size of the `Badge`.
   */
  size?: number,
  /**
   * The content to which the `Badge` is attached.
   */
  children: React.Node,
  style?: any,
  /**
   * @optional
   */
  theme: Theme,
|};

type State = {
  opacity: Animated.Value,
};

/**
 * Badges are small status descriptors for UI elements.
 * A badge consists of a small circle, typically containing a number or other short set of characters, that appears in proximity to another object.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img src="screenshots/badge.png" />
 *     <figcaption>Button with badge</figcaption>
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { Badge, Button } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <Badge value={3}>
 *     <Button mode="outlined" onPress={() => {}}>
 *       Display notifications
 *     </Button>
 *   </Badge>
 * );
 *
 * export default MyComponent;
 * ```
 */
class Badge extends React.Component<Props, State> {
  static defaultProps = {
    verticalPosition: 'top',
    horizontalPosition: 'right',
    size: 12,
  };

  state = {
    opacity: new Animated.Value(this.props.value ? 1 : 0),
  };

  componentDidUpdate(prevProps: Props) {
    const { value } = this.props;

    if (value !== prevProps.value) {
      Animated.timing(this.state.opacity, {
        toValue: value ? 1 : 0,
        duration: 150,
      }).start();
    }
  }

  render() {
    const {
      children,
      horizontalPosition,
      size,
      style,
      theme,
      value,
      verticalPosition,
    } = this.props;
    const { opacity } = this.state;

    const { backgroundColor = theme.colors.badge, ...restStyle } =
      StyleSheet.flatten(style) || {};
    const textColor = color(backgroundColor).dark() ? white : black;

    // $FlowFixMe
    const borderRadius = size / 2;
    // $FlowFixMe
    const paddingHorizontal = size / 2;
    // $FlowFixMe
    const offset = size / -2;

    return (
      <View>
        <View>{children}</View>

        <Animated.View
          style={[
            styles.badge,
            {
              opacity,
              backgroundColor,
              borderRadius,
              paddingHorizontal,
              // $FlowFixMe
              [verticalPosition]: offset,
              // $FlowFixMe
              [horizontalPosition]: offset,
            },
            restStyle,
          ]}
        >
          <Text style={{ color: textColor }}>{value}</Text>
        </Animated.View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  badge: {
    position: 'absolute',
  },
});

export default withTheme(Badge);
