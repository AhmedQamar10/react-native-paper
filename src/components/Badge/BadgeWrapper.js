/* @flow */

import * as React from 'react';
import { View, StyleSheet } from 'react-native';
import Badge from './Badge';

type Props = {|
  /**
   * Value of the `Badge`.
   */
  value?: string | number,
  /**
   * Size of the `Badge`.
   */
  size?: number,
  /**
   * Vertical position of `Badge`.
   */
  verticalPosition?: 'top' | 'bottom',
  /**
   * Horizontal position of `Badge`.
   */
  horizontalPosition?: 'left' | 'right',
  /**
   * The content to which the `Badge` is attached.
   */
  children: React.Node,
  style?: any,
|};

/**
 * BadgeWrapper can be used to wrap an existing UI elements and position the Badge automatically.
 *
 * <div class="screenshots">
 *   <figure>
 *     <img src="screenshots/badge-wrapper.png" />
 *     <figcaption>Button with badge wrapper</figcaption>
 *   </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { Badge, Button } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <Badge.Wrapper value={3}>
 *     <Button mode="outlined" onPress={() => {}}>
 *       Display notifications
 *     </Button>
 *   </Badge.Wrapper>
 * );
 *
 * export default MyComponent;
 * ```
 */
class BadgeWrapper extends React.Component<Props> {
  static displayName = 'Badge.Wrapper';

  static defaultProps = {
    verticalPosition: 'top',
    horizontalPosition: 'right',
    size: 12,
  };

  render() {
    const {
      children,
      horizontalPosition,
      size,
      style,
      value,
      verticalPosition,
    } = this.props;

    // $FlowFixMe
    const offset = size / -2;

    return (
      <View>
        <View>{children}</View>

        <Badge
          size={size}
          style={[
            styles.badge,
            {
              // $FlowFixMe
              [verticalPosition]: offset,
              // $FlowFixMe
              [horizontalPosition]: offset,
            },
            style,
          ]}
        >
          {value}
        </Badge>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  badge: {
    position: 'absolute',
  },
});

export default BadgeWrapper;
