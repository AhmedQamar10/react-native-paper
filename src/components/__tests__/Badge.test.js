/* @flow */

import * as React from 'react';
import renderer from 'react-test-renderer';
import Badge from '../Badge';
import { red500 } from '../../styles/colors';

// Make sure any animation finishes before checking the snapshot results
jest.mock('Animated', () => {
  const ActualAnimated = jest.requireActual('Animated');

  return {
    ...ActualAnimated,
    timing: (value, config) => ({
      start: callback => {
        value.setValue(config.toValue);
        callback && callback({ finished: true });
      },
    }),
  };
});

it('renders no badge', () => {
  // $FlowFixMe
  const tree = renderer.create(<Badge value={null}>Text</Badge>).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders numeric badge', () => {
  const tree = renderer.create(<Badge value={3}>Text</Badge>).toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders badge in top right corner by default', () => {
  const defaultTree = renderer.create(<Badge value="3">Text</Badge>).toJSON();
  const tree = renderer
    .create(
      <Badge value="3" verticalPosition="top" horizontalPosition="right">
        Text
      </Badge>
    )
    .toJSON();

  expect(defaultTree).toEqual(tree);
});

it('renders badge in top right corner', () => {
  const tree = renderer
    .create(
      <Badge value="3" verticalPosition="top" horizontalPosition="right">
        Text
      </Badge>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders badge in bottom right corner', () => {
  const tree = renderer
    .create(
      <Badge value="3" verticalPosition="bottom" horizontalPosition="right">
        Text
      </Badge>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders badge in bottom left corner', () => {
  const tree = renderer
    .create(
      <Badge value="3" verticalPosition="bottom" horizontalPosition="left">
        Text
      </Badge>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders badge in top left corner', () => {
  const tree = renderer
    .create(
      <Badge value="3" verticalPosition="top" horizontalPosition="left">
        Text
      </Badge>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders badge in different color', () => {
  const tree = renderer
    .create(
      <Badge value="3" style={{ backgroundColor: red500 }}>
        Text
      </Badge>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders badge in different size', () => {
  const tree = renderer
    .create(
      <Badge value="3" size={12}>
        Text
      </Badge>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
