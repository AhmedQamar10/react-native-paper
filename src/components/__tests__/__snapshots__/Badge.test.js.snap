// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders badge in bottom left corner 1`] = `
<View>
  <View>
    Text
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 6,
        "bottom": -6,
        "left": -6,
        "opacity": 1,
        "paddingHorizontal": 6,
        "position": "absolute",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Helvetica Neue",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Object {
            "color": "#ffffff",
          },
        ]
      }
    >
      3
    </Text>
  </View>
</View>
`;

exports[`renders badge in bottom right corner 1`] = `
<View>
  <View>
    Text
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 6,
        "bottom": -6,
        "opacity": 1,
        "paddingHorizontal": 6,
        "position": "absolute",
        "right": -6,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Helvetica Neue",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Object {
            "color": "#ffffff",
          },
        ]
      }
    >
      3
    </Text>
  </View>
</View>
`;

exports[`renders badge in different color 1`] = `
<View>
  <View>
    Text
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#e91e63",
        "borderRadius": 6,
        "opacity": 1,
        "paddingHorizontal": 6,
        "position": "absolute",
        "right": -6,
        "top": -6,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Helvetica Neue",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Object {
            "color": "#ffffff",
          },
        ]
      }
    >
      3
    </Text>
  </View>
</View>
`;

exports[`renders badge in different size 1`] = `
<View>
  <View>
    Text
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 6,
        "opacity": 1,
        "paddingHorizontal": 6,
        "position": "absolute",
        "right": -6,
        "top": -6,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Helvetica Neue",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Object {
            "color": "#ffffff",
          },
        ]
      }
    >
      3
    </Text>
  </View>
</View>
`;

exports[`renders badge in top left corner 1`] = `
<View>
  <View>
    Text
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 6,
        "left": -6,
        "opacity": 1,
        "paddingHorizontal": 6,
        "position": "absolute",
        "top": -6,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Helvetica Neue",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Object {
            "color": "#ffffff",
          },
        ]
      }
    >
      3
    </Text>
  </View>
</View>
`;

exports[`renders badge in top right corner 1`] = `
<View>
  <View>
    Text
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 6,
        "opacity": 1,
        "paddingHorizontal": 6,
        "position": "absolute",
        "right": -6,
        "top": -6,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Helvetica Neue",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Object {
            "color": "#ffffff",
          },
        ]
      }
    >
      3
    </Text>
  </View>
</View>
`;

exports[`renders no badge 1`] = `
<View>
  <View>
    Text
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 6,
        "opacity": 0,
        "paddingHorizontal": 6,
        "position": "absolute",
        "right": -6,
        "top": -6,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Helvetica Neue",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Object {
            "color": "#ffffff",
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`renders numeric badge 1`] = `
<View>
  <View>
    Text
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#6200ee",
        "borderRadius": 6,
        "opacity": 1,
        "paddingHorizontal": 6,
        "position": "absolute",
        "right": -6,
        "top": -6,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "Helvetica Neue",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Object {
            "color": "#ffffff",
          },
        ]
      }
    >
      3
    </Text>
  </View>
</View>
`;
